### DISCUSSION TITLE
NewGRF Town Spec ideas

### DISCUSSION THREAD

**Discussion by 2TallTyler** [OpenTTD/OpenTTD#8635] [Features]
There has been some discussion about a future NewGRF town spec, but as I have yet to see any planning documents, I figured I'd start a discussion to collect ideas in one place.

As with the successful discussions about default settings (#8393) and network improvements (#8420), let's post each idea in a separate comment and discuss each idea in sub-comments.

**Comment by 2TallTyler** [comment]
## Property: Road grid spacing

![An example of a wide grid spacing](https://user-images.githubusercontent.com/1780327/68070671-122f6480-fd69-11e9-9366-0dee156295d0.png)

This was proposed as a game setting or GS feature in #7745 and #8442 and rejected in favor of a town NewGRF feature. Read up on this feature in those PRs.

**Reply by ldpl** [reply]
I'm just going to comment on this one but I think this whole discussion is based on a completely wrong presumption. You make it sound like NewGRFs are the way to go and all we need is to decide on the spec for a few properties and callbacks. No, hell no. NewGRFs are affected by all the same issues that got #8442 closed and many-many more. NewGRFs don't know anything about the global state or goal of the game, there is no sane way to do logic in newgrf atm, they're conflicting with each other and so on. In #8442 andythenorth cited industries as an example of a good spec but to me it's all a huge failure because while it may have given us a few good industry sets, I've lost count to how many ideas it made impossible. Some kind of alternative economy or supply-demand simulation? Impossible. Non-random industry placement? Nearly impossible. You basically have to fork FIRS to do anything at all with industries now.

**Reply by LC-Zorg** [reply]
If it were to be an add-on that only determines the introduction of a different road layout (IRL? or similar), replacing the original mechanism, then I'm in favor.
If the new road construction mechanic were to be part of a set of buildings, I am against it - today you can add a lot of building add-ons to the game by mixing them freely. If such an interesting addition would require the use of certain buildings, it would be a bit annoying. ;)
A while back, MagicBuzz created an interesting MagicGS script that builds roads between cities and does it really well (omitting the excessive number of bridges).
{{{The obvious and biggest disadvantage of the scripts is that you can only use one in the game and they cannot be combined. As a result, it often loses to another more important script. NewGRF add-ons do not have this limitation and this is their biggest advantage here. In general, I would be in favor of assigning quests to scripts and allowing multiple scripts to be added to the game at the same time as long as they perform different tasks (city development, industry development, taxes, etc.) If not, I would be in favor of abandoning GS in favor of NewGRF.}}}
It would be good if NewGRF could also build roads between cities. And between metropolises there are dual carriageways. In each case, taking into account the road class.

**Comment by 2TallTyler** [comment]
## Property: Radius of each town zone

It would be nice to control how the town allocates its town zones. At the moment, TOWNZONE_CENTRE has an inordinately large area with all the other zones forming thin bands around the edges.

A good way to set this might be to set each zone as a percentage of the outer radius.

**Reply by LC-Zorg** [reply]
Yes, that would be good, but ...
1. What if you add several NewGRFs with buildings? The last on the list would decide?
2. Maybe it would be worth introducing such modifications in the game itself? Small towns of up to 10,000 inhabitants look relatively good, but above this size the division of the zone disappears and the cities begin to resemble - as someone once neatly put it - "cow poo in the meadow".
3. Perhaps it would be worth introducing something like an "improved city layout" to the game core? It could implement an ITL mechanism or the like, but which would work for each set of buildings? Perhaps the effect could be based on the mechanism of city zones, where, apart from the very center of the city, also near bus stops and certain municipal enterprises (e.g. banks), the zones would be correspondingly higher. This would create the possibility of creating asymmetric districts and cities, also randomly. If it makes sense and interests someone, I could develop this concept.
![City Zones v0 2](https://user-images.githubusercontent.com/67662046/106822935-5a31eb00-6680-11eb-91de-b7ab86093085.png)

**Comment by 2TallTyler** [comment]
## Variable: Date cargo last accepted

This variable would return the date the cargo label passed as an argument was last accepted inside the town. This would be quite useful for houses to query via parent scope, and could also be used by the town's own growth callback.

**Comment by 2TallTyler** [comment]
## Callback: Growth

Called whenever the town wants to grow, and controls whether a town is growing and how quickly. ~~Why let GS have all the fun?~~

(This might be a bad idea. 😛 )

**Reply by LC-Zorg** [reply]
Here I will repeat what I wrote earlier:
{{{The obvious and biggest disadvantage of the scripts is that you can only use one in the game and they cannot be combined. As a result, often one interesting script concerning some smaller element loses to another more important script. NewGRF add-ons do not have this limitation and this is their biggest advantage here. Another very big disadvantage of scripts is that the game ignores them and perceives them as irrelevant for the scenario. Scripts used in online games (the same as those available in online content) do not work after loading a saved game - cities grow endlessly or never start to grow. In general, I would be in favor of assigning quests to scripts and allowing multiple scripts to be added to the game at the same time as long as they perform different tasks (city development, industry development, taxes, etc.) If not, I would be in favor of abandoning GS in favor of NewGRF.}}}

Deciding about growth, introducing new conditions without telling the player what the growth depends on would be very poor. An example is Wasteland, where it is difficult to know what needs to be done to get the right effect. Looking at how players play on servers, few understand the basic mechanics of development. The vast majority of players, especially new ones, believe that the development is influenced by the delivery of goods to cities and the coverage of the largest possible area of ​​the city by transport. It would be very useful to know if the conditions change or complicate. Preferably in the city window.

**Comment by 2TallTyler** [comment]
## Callback: Maximum town bridge length

Control the length limit of bridges constructed by towns. #8439 changed this to be scaled by town population. This callback, if used, would give towns control over the limit. Note that there are separate limits for ramped bridges over rivers and railways, versus flat bridges over ocean.

## Callback: Maximum town tunnel length

The same as above, but for tunnels. Towns gained the ability to build tunnels in #8473, and it is also scaled by town population. There are also separate limits for tunnels through mountains, versus tunnels under obstructions. Note that towns will tunnel under any obstructed tile, including houses blocking its path, if the tunnel would be short enough.

**Reply by LC-Zorg** [reply]
**Maximum town bridge length**
I don't see any particular need for change here. The modification proposed by perezdidac seems very good. Perhaps only the minimum value should be 5 to allow a bridge to be built over the diagonal tracks.
Currently, the problem is not so much the length of the bridge as its location. Very often the city builds bridges along the coasts which seems poor and should not be the case. Building a thousand bridges side by side is also a problem - maybe the number of bridges or the sum of their lengths should be correlated with the size of the city?

**Maximum town tunnel length**
Here, too, I don't see any particular need to modify the values you have set yourself. ;) They seem good and I don't think it's worth surprising the player with changes at every opportunity.

**Comment by zephyris** [comment]
### Variable: x_coordinate
### Variable: y_coordinate
The X and Y coordinates of the town centre tile. Currently this information is available via `0x80`, but wrapped as a confirmed stable variable would make much more sense. Using this variable would allow theming of towns based on position in the map, or perhaps custom town zones by distance calculation from the city centre.

**Comment by zephyris** [comment]
### Variable: terrain_type
Tile type of the town centre tile. Useful for theming of towns based on their climate. Also useful for towns being aware of being above the snow line for town growth/requiring food.
### Variable: tile_height
Tile height of the town centre. Might as well expose this simple variable, might be useful for giving a town a coastal vs alpine appearance, or similar.

**Comment by zephyris** [comment]
### Variable: nearest_tile_matching_criterion
Similar to `nearest_house_matching_criterion` for houses, this would take a radius (1..63) parameter and search for then return the distance to the nearest tile to the city centre matching a particular tile class. Could also be extended, using a second parameter, to search for the distance to the nearest tile class, tile terrain type, water tile class, and anything else I'm forgetting. Useful for detecting if a town is a riverside town, a town with a canal, a town near the coast, etc. for theming it.

**Reply by zephyris** [reply]
### Variable: nearby_tile_height, nearby_tile_class, nearby_tile_slope, etc.
These would take X- and Y-offsets (-8..7), and return if that tile position relative to the city centre matches the criterion. I'm less sure of the utility of these compared to `nearest_tile_matching_criterion`, perhaps `nearby_tile_height` and `nearby_tile_slope` could be used to detect a town in hilly terrain?

**Comment by zephyris** [comment]
### Variable: Cargo accepted this month
### Variable: Cargo accepted last month
### Variable: Cargo transported this month
### Variable: Cargo transported last month
These variables would return the quantity of cargo, using a cargo label passed as an argument, transported or produced by the town. This allows visual feedback for delivery of passengers, mail, goods or food in building appearance. It'd also enable some industry-like behaviour for houses to produce cargo by callback in response to cargo delivered to the town, eg. produce goods in response to quantity of goods delivered.

**Comment by zephyris** [comment]
### Variable: Cargo percent transported this month
### Variable: Cargo percent transported last month
These variables would return the percentage of cargo produced by the town, using a cargo label passed as an argument, that was actually transported. This effectively generalises `percent_transported_passengers` and `percent_transported_mail`.

**Comment by zephyris** [comment]
### Callback: name
Set the name of the town. Could simply point to a string identifier, but would be _really_ powerful if it could point to a top-level `town_names` block. ie. NML town name generation.

**Comment by zephyris** [comment]
### Callback: street furniture
Points to a sprite set which defines street furniture (ie. replace the street light or roadside tree) sprites. Pointing to a 1-member set could give orientationless sprites (like trees), a 2-member set  could give W/E oriented sprites (like street lights), a 4-member set for SE, SW, NE, NW oriented sprites. Would probably need to have multiple callbacks, one per town zone.

**Comment by andythenorth** [comment]
Variable: does town have company statue(s).

This might not be possible as we might not have any way to get the company context.
